Fix inline compilation benchmarks and Improve searching Graph.Link performance

hubertp opened the pull request:
### Pull Request Description

As benchmarks show, a significant amount of time is spent traversing `Set` of `Graph.Link`s. That's unfortunate and unnecessary. We can equally keep helper maps that make search constant time.

Fixed inline compilation benchmarks by properly cleaning up scopes after runs.

Closes #9237.

### Important Notes

Things like
![Screenshot from 2024-03-22 11-23-01](https://github.com/enso-org/enso/assets/292128/7c1e220a-6e33-4396-a9b2-0e788f615323)
![Screenshot from 2024-03-22 11-13-19](https://github.com/enso-org/enso/assets/292128/0272b1cb-252e-4662-b539-174844941c8e)
are all gone. There is plenty of it those are just samples.

Benchmarks are back in order:
```
[info] # Warmup Iteration   1: 2.702 ms/op
[info] # Warmup Iteration   2: 3.080 ms/op
[info] # Warmup Iteration   3: 2.818 ms/op
[info] # Warmup Iteration   4: 3.334 ms/op
[info] # Warmup Iteration   5: 2.448 ms/op
[info] # Warmup Iteration   6: 2.583 ms/op
[info] Iteration   1: 2.908 ms/op
[info] Iteration   2: 2.915 ms/op
[info] Iteration   3: 2.774 ms/op
[info] Iteration   4: 2.601 ms/op
[info] Result "org.enso.compiler.benchmarks.inline.InlineCompilerBenchmark.longExpression":
[info]   2.799 ±(99.9%) 0.953 ms/op [Average]
[info]   (min, avg, max) = (2.601, 2.799, 2.915), stdev = 0.148
[info]   CI (99.9%): [1.846, 3.753] (assumes normal distribution)
```

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Improve performance for searching Graph.Link

As benchmarks show, a significant amount of time is spent traversing
`Set` of `Graph.Link`s. That's unfortunate and unnecessary. We can
equally keep helper maps that make search constant time.

hubertp committed:
nit

4e6 commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/Graph.scala:
```diff
@@ -5,14 +5,17 @@ import org.enso.syntax.text.Debug
+  var rootScope: Graph.Scope                              = new Graph.Scope()
+  private var links: Set[Graph.Link]                      = Set()
+  private var sourceLinks: Map[Graph.Id, Set[Graph.Link]] = new HashMap()
+  private var targetLinks: Map[Graph.Id, Set[Graph.Link]] = new HashMap()
```
There is an implementation of serialization logic for the `Graph` https://github.com/enso-org/enso/blob/6665c22eb9d5d8b1fd16ead769acaa3c31dda57b/engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/PassPersistance.java#L153-L157

Please make sure that it works correctly with the newly added fields and probably update the serialization id

hubertp replied:
Should work correctly as I'm reproducing `sourceLinks` and `targetLinks` from `links` when deserialized. I'm not serializing those fields as there is no need for that (`links` continues to be serialized).

hubertp committed:
adapt tests

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/Graph.scala:
```diff
@@ -5,14 +5,17 @@ import org.enso.syntax.text.Debug
-  var rootScope: Graph.Scope = new Graph.Scope()
-  var links: Set[Graph.Link] = Set()
-  var nextIdCounter          = 0
+  var rootScope: Graph.Scope                              = new Graph.Scope()
```
If `initXyz` is better pattern than calling `g.rootScope_$eq(rootScope);` - then `rootScope` could be made private as well, possibly.

hubertp replied:
I tried and looks like a lot more stuff depends on it being public. I tried to keep the PR succinct for now.

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/Graph.scala:
```diff
@@ -24,11 +27,26 @@ sealed class Graph extends Serializable {
     copy
   }
 
+  def initLinks(links: Set[Graph.Link]): Unit = {
```
Shouldn't `initLinks` also populate `getLinks`? I don't see such behavior and I believe that calling:
```scala
val set = Set(l1, l2, l3);
g.initLinks();
assert g.getLinks() == set
```
would fail.

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/Graph.scala:
```diff
@@ -24,11 +27,26 @@ sealed class Graph extends Serializable {
   }
 
+  def initLinks(links: Set[Graph.Link]): Unit = {
+    links.foreach { link =>
```
Shouldn't `sourceLinks` and `targetLinks` be cleared first? Otherwise calling `initLinks` multiple times will accumulate the results. Either:
- allow `initLinks` to be called only once
- clear the content if called multiple times
- rename to `addLinks`

hubertp replied:
Yes, if you call it multiple times that will be the case. I guess clearing things sounds OK.

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/Graph.scala:
```diff
@@ -24,11 +27,26 @@ sealed class Graph extends Serializable {
+  def initLinks(links: Set[Graph.Link]): Unit = {
+    links.foreach { link =>
+      sourceLinks = sourceLinks.updatedWith(link.source)(v =>
+        v.map(s => s + link).orElse(Some(Set(link)))
```
According to [scaladoc](https://www.scala-lang.org/api/2.13.7/scala/collection/immutable/Map.html#updatedWith[V1%3E:V](key:K)(remappingFunction:Option[V]=%3EOption[V1]):CC[K,V1]) this is going to add to the `Set` associated with `link.source`. _Just a note_ to be able to read what's happening.

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/Graph.scala:
```diff
@@ -129,7 +155,10 @@ sealed class Graph extends Serializable {
     */
   def linksFor(id: Graph.Id): Set[Graph.Link] = {
-    links.filter(l => l.source == id || l.target == id)
+    sourceLinks.getOrElse(id, Set.empty[Graph.Link]) ++ targetLinks.getOrElse(
```
Great finding. Good fix. Thank you.

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/Graph.scala:
```diff
@@ -85,6 +105,12 @@ sealed class Graph extends Serializable {
   ): Option[Graph.Link] = {
     scopeFor(occurrence.id).flatMap(_.resolveUsage(occurrence).map { link =>
       links += link
+      sourceLinks = sourceLinks.updatedWith(link.source)(v =>
```
Looks like the same code as in `initLinks`. Maybe really rename `initLinks` to `addLinks` and call it here?

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/Graph.scala:
```diff
@@ -24,11 +27,26 @@ sealed class Graph extends Serializable {
+    }
+  }
+
+  def getLinks(): Set[Graph.Link] = links
```
Wouldn't it be better to remove the `links` field all together and just compute the links from `sourceLinks` and `targetLinks`? The logic between `initLinks` and `getLinks` & co. would be simpler and possibly `getLinks()` isn't on critical path. Otherwise the [inconsistency](https://github.com/enso-org/enso/pull/9520#discussion_r1537195157) must have been caught somewhere, right?

hubertp replied:
I kind of like that `links` is the source of truth and `sourceLinks/targetLinks` serve as aux indexes for quicker access. It also simplifies serde.

JaroslavTulach reviewed (approved):
Glad to see a speedup. Some comments related to the code left inline.

hubertp committed:
PR review

hubertp committed:
Fix inline benchmarks

Treat InlineContext as a closeable resource, to ensure that benchmarks
properly cleanup.
The problem was that the number of scopes has been steadily increasing
with every benchmark run. Traversing such additional scopes took time,
hence exponential blow up in run times.

hubertp commented:
With the last change the inline compilation benchmarks are back in order:
```
[info] # Warmup Iteration   1: 2.702 ms/op
[info] # Warmup Iteration   2: 3.080 ms/op
[info] # Warmup Iteration   3: 2.818 ms/op
[info] # Warmup Iteration   4: 3.334 ms/op
[info] # Warmup Iteration   5: 2.448 ms/op
[info] # Warmup Iteration   6: 2.583 ms/op
[info] Iteration   1: 2.908 ms/op
[info] Iteration   2: 2.915 ms/op
[info] Iteration   3: 2.774 ms/op
[info] Iteration   4: 2.601 ms/op
[info] Result "org.enso.compiler.benchmarks.inline.InlineCompilerBenchmark.longExpression":
[info]   2.799 ±(99.9%) 0.953 ms/op [Average]
[info]   (min, avg, max) = (2.601, 2.799, 2.915), stdev = 0.148
[info]   CI (99.9%): [1.846, 3.753] (assumes normal distribution)
```

JaroslavTulach commented on engine/runtime-benchmarks/src/main/java/org/enso/compiler/benchmarks/inline/InlineContextResource.java:
```diff
@@ -0,0 +1,24 @@
+        .foreach(
+            s -> {
+              s.scope().close();
+              return BoxedUnit.UNIT;
```
Please simplify and use `null`. That will remove needless dependency on Scala core libraries.

JaroslavTulach commented on engine/runtime-benchmarks/src/main/java/org/enso/compiler/benchmarks/inline/InlineContextResourceBuilder.java:
```diff
@@ -0,0 +1,28 @@
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.scope.ModuleScope;
+
+public record InlineContextResourceBuilder(
```
Builder pattern has setter methods that return `this` and mutate its internal state. According to the previous description `record` cannot be a _builder_ - it might be a [cumulative factory](http://wiki.apidesign.org/wiki/CumulativeFactory), however.

hubertp replied:
Fair enough, thanks

JaroslavTulach commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/alias/Graph.scala:
```diff
@@ -675,6 +675,17 @@ object Graph {
     }
+
+    private def removeScope(scope: Scope): Unit = {
+      childScopes = childScopes.filter(_ != scope)
```
Just a _threading note_: this is probably not thread safe... in case we want to evaluate _visualizations_ in parallel, we may face issues like this at various places.

hubertp replied:
Yeah, we are likely to open a can of worms.

hubertp committed:
nits

hubertp committed:
Merge branch 'develop' into wip/hubert/compiler-inline-perf
